float32 r_ref
float32 l_ref
